File: prototype/rails/src/app/actions/customers/resolve_customer_from_cookie.rb
module Customers
  class ResolveCustomerFromCookie
    # Returns nil when the cookie is absent, unreadable, or points at a customer
    # that no longer exists.
    def call(cookie_value)
      id = Integer(cookie_value, exception: false)
      return nil if id.nil? || id < 1

      Customer.find_by(id: CustomerMerge.where(anonymous_customer_id: id).pick(:customer_id) || id)
    end
  end
end

File: prototype/php/src/app/Actions/Customers/ResolveCustomerFromCookie.php
<?php

declare(strict_types=1);

namespace App\Actions\Customers;

use App\Models\Customer;
use App\Models\CustomerMerge;

final readonly class ResolveCustomerFromCookie
{
    /**
     * @return Customer|null null when the cookie is absent, unreadable, or
     *                       points at a customer that no longer exists
     */
    public function __invoke(?string $cookieValue): ?Customer
    {
        $id = filter_var($cookieValue, FILTER_VALIDATE_INT);

        if ($id === false || $id < 1) {
            return null;
        }

        $mergedInto = CustomerMerge::where('anonymous_customer_id', $id)->value('customer_id');
        $mergedIntoId = filter_var($mergedInto, FILTER_VALIDATE_INT);

        return Customer::find($mergedIntoId === false ? $id : $mergedIntoId);
    }
}
